examples/door-lock-app/linux/include/LockEndpoint.h

```diff
@@ -0,0 +1,112 @@
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/clusters/door-lock-server/door-lock-server.h>
+#include <vector>
+
+struct LockUserInfo;
+struct LockCredentialInfo;
+struct WeekDaysScheduleInfo;
+struct YearDayScheduleInfo;
+
+static constexpr size_t DOOR_LOCK_CREDENTIAL_INFO_MAX_DATA_SIZE = 20;
+
+class LockEndpoint
+{
+public:
+    LockEndpoint(chip::EndpointId endpointId, uint16_t numberOfLockUsersSupported, uint16_t numberOfCredentialsSupported,
+                 uint8_t weekDaySchedulesPerUser, uint8_t yearDaySchedulesPerUser) :
+        mEndpointId(endpointId),
+        mLockState(DlLockState::kLocked), mLockUsers(numberOfLockUsersSupported),
+        mLockCredentials(numberOfCredentialsSupported + 1),
+        mWeekDaySchedules(numberOfLockUsersSupported, std::vector<WeekDaysScheduleInfo>(weekDaySchedulesPerUser)),
+        mYearDaySchedules(numberOfLockUsersSupported, std::vector<YearDayScheduleInfo>(numberOfLockUsersSupported))
+    {}
+
+    inline chip::EndpointId GetEndpointId() { return mEndpointId; }
+
+    bool Lock(chip::Optional<chip::ByteSpan> pin);
+    bool Unlock(chip::Optional<chip::ByteSpan> pin);
+
+    bool GetUser(uint16_t userIndex, EmberAfPluginDoorLockUserInfo & user) const;
+    bool SetUser(uint16_t userIndex, chip::FabricIndex creator, chip::FabricIndex modifier, const chip::CharSpan & userName,
+                 uint32_t uniqueId, DlUserStatus userStatus, DlUserType usertype, DlCredentialRule credentialRule,
+                 const DlCredential * credentials, size_t totalCredentials);
+
+    bool GetCredential(uint16_t credentialIndex, DlCredentialType credentialType,
+                       EmberAfPluginDoorLockCredentialInfo & credential) const;
+
+    bool SetCredential(uint16_t credentialIndex, DlCredentialStatus credentialStatus, DlCredentialType credentialType,
+                       const chip::ByteSpan & credentialData);
+
+    DlStatus GetSchedule(uint8_t weekDayIndex, uint16_t userIndex, EmberAfPluginDoorLockWeekDaySchedule & schedule);
+    DlStatus GetSchedule(uint8_t yearDayIndex, uint16_t userIndex, EmberAfPluginDoorLockYearDaySchedule & schedule);
+    DlStatus SetSchedule(uint8_t weekDayIndex, uint16_t userIndex, DlScheduleStatus status, DlDaysMaskMap daysMask,
+                         uint8_t startHour, uint8_t startMinute, uint8_t endHour, uint8_t endMinute);
+    DlStatus SetSchedule(uint8_t yearDayIndex, uint16_t userIndex, DlScheduleStatus status, uint32_t localStartTime,
+                         uint32_t localEndTime);
+
+private:
+    bool setLockState(DlLockState lockState, chip::Optional<chip::ByteSpan> & pin);
+    const char * lockStateToString(DlLockState lockState) const;
+
+    chip::EndpointId mEndpointId;
+    DlLockState mLockState;
+
+    // This is very naive implementation of users/credentials/schedules database and by no means the best practice. Proper storage
+    // of those items is out of scope of this example.
```

Comment:
Seems like the right thing is to have the storage app-managed in some way but the logic shared across all apps using this cluster and all platforms....  Followup?

Reply:
I'm not sure what you mean. The original idea is to have a door lock cluster free of any storage and let the app provide all the data via callbacks such as `emberAfPluginDoorLockGetSchedule` and `emberAfPluginDoorLockSetSchedule`. The cluster code in the [door-lock-server.cpp](https://github.com/project-chip/connectedhomeip/pull/13863/files/142100859ff50d0deef3e038d3201ed795a64522#diff-ba4731e35f6cf29195b5ca8a6c50f6ed2b1730e9c492ae6453d02ac34e9f0f74) only calls those callbacks to retrieve/set stuff and it's up to the app to implement actual storage. Sure, the [door-lock-server.cpp](https://github.com/project-chip/connectedhomeip/pull/13863/files/142100859ff50d0deef3e038d3201ed795a64522#diff-ba4731e35f6cf29195b5ca8a6c50f6ed2b1730e9c492ae6453d02ac34e9f0f74) also has those callbacks [defined](https://github.com/project-chip/connectedhomeip/pull/13863/files/142100859ff50d0deef3e038d3201ed795a64522#diff-ba4731e35f6cf29195b5ca8a6c50f6ed2b1730e9c492ae6453d02ac34e9f0f74R3091) but they are marked weak and they do nothing.  The `LockEndpoint.h` is part of the example application (examples/door-lock-app) and the storage logic is implemented only there.

Reply:
I guess I was trying to figure out whether any of this logic would be common to all database impls or not.  Sounds like not.